R/utils.R

```diff
@@ -0,0 +1,146 @@
+# Implements "whitespace eating" functionality on top of a normal TextWriter.
+# (WSTextWriter doesn't inherit from TextWriter, as it presents a different
+# interface.)
+#
+# WSTextWriter relies on the caller distinguishing between writes of important
+# content, and writes of whitespace that may or may not be elided (`.$write()`
+# vs `.$writeWS()`).
+#
+# At any point, `eatWS` may be called, which will cause any recent `writeWS`
+# operations (i.e. those since either the beginning of time, or the most recent
+# `write` operation) to be undone, AND for any future `writeWS` calls to be
+# ignored. A call to `write` will be respected, and will restore normal
+# behavior.
+WSTextWriter <- R6Class("WSTextWriter",
+  private = list(
+    writer = "TextWriter"
+  ),
+  public = list(
+    initialize = function() {
+      private$writer <- TextWriter$new()
+    },
+    close = function() {
+      private$writer$close()
+    },
+    write = function(text) {
+      # If an error is going to happen while evaluating `text`, let's make it
+      # happen before we mutate any of our internal state
+      force(text)
+
+      # Stop eating whitespace
+      private$writer$setSuppress(FALSE)
+
+      private$writer$write(text)
+
+      # Set a bookmark here, so that future calls to eatWS() don't affect this
+      # write
+      private$writer$savePosition()
+    },
+    # Write whitespace. If suppressWhitespace() was called and
+    # its effect has not been canceled, then this method no-ops.
+    # @param text Single element character vector containing only
+    #   whitespace characters
+    writeWS = function(text) {
```

Comment:
Should the "whitespace only" expectation be enforced? I realize why we would opt not to for performance, but I'm wondering if it's desirable.